File: 엄마호랑이반/9주차 문제/이다영/등굣길.java
package temp_dev;

import java.util.*;
import java.io.*;

public class Programmers_2 {
	public static void main(String[] args) {
		System.out.println(solution(4, 3, new int[][] {{2,2}}));
		
	}
	
	// 효율성 안됨 망알
	public static int solution(int m, int n, int[][] puddles) {
		int[][] dp = new int[n + 1][m + 1];
		dp[1][1] = 1;
		
		
		for (int[] puddle : puddles) {
			dp[puddle[1]][puddle[0]] = -1;
		}
		
		for (int x = 1; x < n + 1; x++) {
			for (int y = 1; y < m + 1; y++) {
				
				if(dp[x][y] != -1) {
					dp[x][y] += dp[x-1][y] + dp[x][y-1];
					
				} else {
					dp[x][y] = 0;
				}
			}
		}
		return dp[n][m] % 1_000_000_007;
    }

}

File: 엄마호랑이반/9주차 문제/박정환/등굣길.py
def solution(m, n, puddles):
    board = [[0] * (m + 1) for _ in range(n+1)]
    board[1][1] = 1

    for y, x in puddles:
        board[x][y] = -1
    
    # print(board)

    for i in range(1, n+1):
        for j in range(1, m+1):
            if i == 1 and j == 1:
                continue
            if board[i][j] == -1:
                board[i][j] = 0
            else:
                board[i][j] = board[i][j-1] + board[i-1][j]
    return board[n][m] % 1000000007


print(solution(4, 3, [[2,2]]))
